packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -113,6 +116,22 @@ class UpgradeCommandRunner {
     @required bool testFlow,
     @required bool verifyOnly,
   }) async {
+    if (flutterVersion.repositoryUrl != _flutterGit) {
+      // If the upstream remote url is configured to track a remote different
+      // from _flutterGit, "flutter upgrade" would not necessarily fetch the
+      // actual upstream updates from the latter, which is used for version
+      // check and inform users about the update in the first place.
+      throwToolExit(
+        'Your local copy of Flutter is tracking an unsupported remote '
+        '"${flutterVersion.repositoryUrl}".\n'
+        'To use the unofficial remote, set the environment variable '
+        '"FLUTTER_GIT_URL" to "${flutterVersion.repositoryUrl}",\n'
+        'or to use the official remote, run '
+        '"git remote add origin https://github.com/flutter/flutter" and '
+        '"git branch --set-upstream-to=origin/${flutterVersion.channel}" if remote "origin" '
```

Comment:
These instructions are a little confusing. If:

1) they have a different `origin` remote configured (which is most likely), they will have to remove that `origin` FIRST, then follow these instructions.
2) They correctly have _flutterGit configured as origin, however the current branch is tracking a different remote. In this case `git branch --set-upstream-to` should be enough to fix.
3) They do not have an `origin` remote configured at all (not sure how they'd get in this state), in which case these instructions would be sufficient.

Because there are these three permutations (and probably more), rather than try to cover every possible state a user could be in and help them recover, maybe we should instead direct them to re-install the SDK with a [link to the website](https://flutter.dev/docs/get-started/install).

Developers who intentionally got into this state via git will know what they're doing and ignore our warning anyway :). They can still do a `git pull` on whatever branch they're actually tracking.

Reply:
> Because there are these three permutations (and probably more)

You mean these? (origin => `github.com/flutter/flutter`, and `FLUTTER_GIT_URL` has a different url than the origin)
|#|Tracking remote|HEAD is pointing to branch?|`_flutterGit`|Result|`git` fetches tags from|
|--:|----------------|-------------------|-----------|-------|--------------------|
|1|origin|Yes|github.com/flutter/flutter|upgrades|tracking remote|
|2|notOrigin|Yes|github.com/flutter/flutter|ToolExit: Your local copy of Flutter is tracking a nonstandard remote|tracking remote|
|3|nothing|Yes|github.com/flutter/flutter|ToolExit: Unable to upgrade Flutter: no origin repository configured.|`origin` remote if it exists|
|---|||||
|4|origin|No|github.com/flutter/flutter|ToolExit: You are not currently on a release branch|tracking remote|
|5|notOrigin|No|github.com/flutter/flutter|ToolExit: You are not currently on a release branch|tracking remote|
|6|nothing|No|github.com/flutter/flutter|ToolExit: You are not currently on a release branch|`origin` remote if it exists|
|---|||||
|7|origin|Yes|`FLUTTER_GIT_URL`|ToolExit: Your local copy of Flutter is tracking a nonstandard remote|tracking remote|
|8|notOrigin|Yes|`FLUTTER_GIT_URL`|upgrades|tracking remote|
|9|nothing|Yes|`FLUTTER_GIT_URL`|ToolExit: Unable to upgrade Flutter: no origin repository configured.|`origin` remote if exists|
|---|||||
|10|origin|No|`FLUTTER_GIT_URL`|ToolExit: You are not currently on a release branch|tracking remote|
|11|notOrigin|No|`FLUTTER_GIT_URL`|ToolExit: You are not currently on a release branch|tracking remote|
|12|nothing|No|`FLUTTER_GIT_URL`|ToolExit: You are not currently on a release branch|`origin` remote if it exists|

> These instructions are a little confusing

I somewhat agree. Case in point `#7` would produce an error something like this:
```
Your local copy of Flutter is tracking a nonstandard remote "https://github.com/flutter/flutter".
To use the unofficial remote, set the environment variable "FLUTTER_GIT_URL" to "https://github.com/flutter/flutter", 
or to use the official remote, run "git remote add origin https://github.com/flutter/flutter" and "git branch --set-upstream-to=origin/dummy_branch" if remote "origin" exists in <working_directory>.
```
The message could be reworded to more clearly inform the users that only `#1` and `#8` are allowed by `flutter upgrade`, for the lack of a better word.

> they have a different origin remote configured (which is most likely)

Flutter can only be obtained from either cloning this repository, or getting an archive from the website, which should have the `origin` remote set properly(can't say about the snap package). I would argue that users who even then change the url of the remote should be able to figure out and delete/set-url of `origin` once they see the error git produces.

> maybe we should instead direct them to re-install the SDK with a link to the website.

This might be helpful for someone who is too scared to work with `git`.
```
As a last resort, if you are OK WITH LOSING ANY LOCAL CHANGES, you can re-install a fresh copy from https://flutter.dev/docs/get-started/install
```

The question is should this be included in every `throwToolExit` message(even those which are not affected by this PR) in `upgrade.dart`?

Reply:
@christopherfujino how about
```
if (flutterGit == https://github.com/flutter/flutter) {
  // FLUTTER_GIT_URL is unset
  Unable to upgrade flutter: The Flutter SDK is tracking a nonstandard remote
  <nonstandard_remote>.
   - To use the current remote, set the environment variable FLUTTER_GIT_URL to
 <nonstandard_remote>, and retry.
   - To use the standard remote, run
        "git remote set-url ____ https://github.com/flutter/flutter", and retry.
        (How do I get the remote's name without duplicating code from version.dart?) 
} else {
  // FLUTTER_GIT_URL is set
  Unable to upgrade flutter: The Flutter SDK is tracking <remote_A> but
  FLUTTER_GIT_URL is set to <remote_B>.
  Either remove FLUTTER_GIT_URL from the environment or set FLUTTER_GIT_URL as
  <remote_A>, and retry.
  // In case they happen to track a nonstandard remote and remove
  // FLUTTER_GIT_URL, upgrading flutter would bring them to the block above.
}
```